### PR TITLE
Run cudagraphs on AOTAutograd cache hit

### PR DESCRIPTION
**Pull request by jamesjwu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132294

This threads through all of the necessary parts into aot autograd from the FXGraphCache changes so that we can run cudagraphs properly on a AOTAutograd cache hit. 

Specifics:
- AOTAutograd needs access to the `cudagraphs` boxedbool in order to properly set the backward to not use cudagraphs on a cache hit from the forward. 
- We have lots of tests that test this already from the previous PR, so I just added an extra test and made the previous test work with both AOTAutogradCache and FXGraphCache at the same time. 

```
TORCH_LOGS=torch._functorch._aot_autograd.autograd_cache,cudagraphs ENABLE_AOT_AUTOGRAD_CACHE=1 TORCHINDUCTOR_FX_GRAPH_CACHE=1 tlp python benchmarks/gpt_fast/benchmark.py --output ~/gpt_fast_benchmark.csv
```
Twice, once on cache miss and once and cache hit. 

Here is the perfetto trace for each(FB only link):

**Cache Miss:**
Logs:
```
Loading model Llama-2-7b-chat-hf
Time to load model: 0.66 seconds
I0813 10:53:34.416000 911030 torch/_functorch/_aot_autograd/autograd_cache.py:479] [0/0] AOTAutograd cache miss for key alqchc7zw6ynsxj2bzktcsngu4cajwcb3tmhvwlyqkuinx3zhmey
I0813 10:53:51.395000 911030 torch/_functorch/_aot_autograd/autograd_cache.py:558] [0/0] Writing AOTAutograd cache entry to /tmp/torchinductor_jjwu/aotautograd/alqchc7zw6ynsxj2bzktcsngu4cajwcb3tmhvwlyqkuinx3zhmey/entry
I0813 10:54:17.579000 911030 torch/_functorch/_aot_autograd/autograd_cache.py:479] [1/0] AOTAutograd cache miss for key a3nq2ywjxku342c6ag7rsqkalnxfshlcgve3tb2bigg7a45uz6pt
I0813 10:54:38.636000 911030 torch/_functorch/_aot_autograd/autograd_cache.py:558] [1/0] Writing AOTAutograd cache entry to /tmp/torchinductor_jjwu/aotautograd/a3nq2ywjxku342c6ag7rsqkalnxfshlcgve3tb2bigg7a45uz6pt/entry
I0813 10:54:39.228000 911030 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 10:54:39.939000 911030 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 0
V0813 10:55:10.615000 911030 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 0 of graph recording id 0
Compilation time: 101.24 seconds
Average tokens/sec: 147.96 tokens/sec
Average bandwidth achieved: 1955.22 GB/s
Memory used: 14.51 GB
```

Chromium Event(fb only):
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom%2Fchromium_events.json#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom%2Fchromium_events.json&local_cache_key

image

**Cache Hit:**
Logs:
```
Loading model Llama-2-7b-chat-hf
Time to load model: 0.67 seconds
I0813 10:55:51.821000 944420 torch/_functorch/_aot_autograd/autograd_cache.py:474] [0/0] AOTAutograd cache hit for key alqchc7zw6ynsxj2bzktcsngu4cajwcb3tmhvwlyqkuinx3zhmey
I0813 10:55:55.465000 944420 torch/_functorch/_aot_autograd/autograd_cache.py:474] [1/0] AOTAutograd cache hit for key a3nq2ywjxku342c6ag7rsqkalnxfshlcgve3tb2bigg7a45uz6pt
I0813 10:55:56.030000 944420 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 10:55:56.192000 944420 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 0
V0813 10:55:56.426000 944420 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 0 of graph recording id 0
Compilation time: 9.40 seconds
Average tokens/sec: 147.94 tokens/sec
Average bandwidth achieved: 1954.98 GB/s
Memory used: 14.51 GB
```
Chromium Event(fb only):
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom2%2Fchromium_events.json#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom2%2Fchromium_events.json&local_cache_key

image

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by bdhirsh** [comment]
> Meta internal tlparse link showing a cache miss and cache hit of nanogpt running AOTAutograd:

> https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpVi1FIl/dedicated_log_torch_trace_bb4zl_bc.log/index.html

How can I tell from that tlparse that we had an AOT cache hit / cache miss (due to cudagraphs)?

**Comment by eellison** [comment]
> Meta internal tlparse link showing a cache miss and cache hit of nanogpt running AOTAutograd:

Sorry would you explain more what is happening here ?

**Comment by jamesjwu** [comment]
Apologies, I never finished writing the description lol. I decided to get the benchmark working for gpt_fast/benchmarks.py instead, so I ran:

```
TORCH_LOGS=torch._functorch._aot_autograd.autograd_cache,cudagraphs ENABLE_AOT_AUTOGRAD_CACHE=1 TORCHINDUCTOR_FX_GRAPH_CACHE=1 tlp python benchmarks/gpt_fast/benchmark.py --output ~/gpt_fast_benchmark.csv
```
Twice, once on cache miss and once and cache hit. 

Here is the perfetto trace for each(FB only link):

**Cache Miss:**
Logs:
```
Loading model Llama-2-7b-chat-hf
Time to load model: 0.66 seconds
I0813 10:53:34.416000 911030 torch/_functorch/_aot_autograd/autograd_cache.py:479] [0/0] AOTAutograd cache miss for key alqchc7zw6ynsxj2bzktcsngu4cajwcb3tmhvwlyqkuinx3zhmey
I0813 10:53:51.395000 911030 torch/_functorch/_aot_autograd/autograd_cache.py:558] [0/0] Writing AOTAutograd cache entry to /tmp/torchinductor_jjwu/aotautograd/alqchc7zw6ynsxj2bzktcsngu4cajwcb3tmhvwlyqkuinx3zhmey/entry
I0813 10:54:17.579000 911030 torch/_functorch/_aot_autograd/autograd_cache.py:479] [1/0] AOTAutograd cache miss for key a3nq2ywjxku342c6ag7rsqkalnxfshlcgve3tb2bigg7a45uz6pt
I0813 10:54:38.636000 911030 torch/_functorch/_aot_autograd/autograd_cache.py:558] [1/0] Writing AOTAutograd cache entry to /tmp/torchinductor_jjwu/aotautograd/a3nq2ywjxku342c6ag7rsqkalnxfshlcgve3tb2bigg7a45uz6pt/entry
I0813 10:54:39.228000 911030 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 10:54:39.939000 911030 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 0
V0813 10:55:10.615000 911030 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 0 of graph recording id 0
Compilation time: 101.24 seconds
Average tokens/sec: 147.96 tokens/sec
Average bandwidth achieved: 1955.22 GB/s
Memory used: 14.51 GB
```

Chromium Event(fb only):
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom%2Fchromium_events.json#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom%2Fchromium_events.json&local_cache_key

image


**Cache Hit:**
Logs:
```
Loading model Llama-2-7b-chat-hf
Time to load model: 0.67 seconds
I0813 10:55:51.821000 944420 torch/_functorch/_aot_autograd/autograd_cache.py:474] [0/0] AOTAutograd cache hit for key alqchc7zw6ynsxj2bzktcsngu4cajwcb3tmhvwlyqkuinx3zhmey
I0813 10:55:55.465000 944420 torch/_functorch/_aot_autograd/autograd_cache.py:474] [1/0] AOTAutograd cache hit for key a3nq2ywjxku342c6ag7rsqkalnxfshlcgve3tb2bigg7a45uz6pt
I0813 10:55:56.030000 944420 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 10:55:56.192000 944420 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 0
V0813 10:55:56.426000 944420 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 0 of graph recording id 0
Compilation time: 9.40 seconds
Average tokens/sec: 147.94 tokens/sec
Average bandwidth achieved: 1954.98 GB/s
Memory used: 14.51 GB
```
Chromium Event(fb only):
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom2%2Fchromium_events.json#!/viewer?url=https%3A%2F%2Finterncache-all.fbcdn.net%2Fmanifold%2Ftlparse_reports%2Ftree%2Flogs%2Fjjwu%2Fcustom2%2Fchromium_events.json&local_cache_key

image

**Comment by jamesjwu** [comment]
And here's the same thing for nanogpt, just for completeness:

```
TORCH_LOGS=torch._functorch._aot_autograd.autograd_cache,cudagraphs ENABLE_AOT_AUTOGRAD_CACHE=1 TORCHINDUCTOR_FX_GRAPH_CACHE=1 tlp benchmarks/dynamo/torchbench.py --training --backend=inductor --only nanogpt --warm-start-latency --performance
Performing cold-start run for nanogpt
loading model: 0it [00:00, ?it/s]number of parameters: 123.69M
num decayed parameter tensors: 50, with 124,354,560 parameters
num non-decayed parameter tensors: 98, with 121,344 parameters
using fused AdamW: True
loading model: 0it [00:03, ?it/s]
cuda train nanogpt
I0813 11:03:02.344000 1005502 torch/_functorch/_aot_autograd/autograd_cache.py:479] [2/0_1] AOTAutograd cache miss for key adctpnaix4qr2ofjsrnfduh5v4m7qsmnnuy532v45igmy2cplgyu
I0813 11:03:10.063000 1005502 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 11:03:10.792000 1005502 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 0
I0813 11:03:14.803000 1005502 torch/_functorch/_aot_autograd/autograd_cache.py:558] [2/0_1] Writing AOTAutograd cache entry to /tmp/tmpxrdfdu4m/aotautograd/adctpnaix4qr2ofjsrnfduh5v4m7qsmnnuy532v45igmy2cplgyu/entry
I0813 11:03:14.803000 1005502 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 11:03:14.805000 1005502 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 1
W0813 11:03:15.654000 1005502 torch/_logging/_internal.py:1058] [5/0] Profiler function <class 'torch.autograd.profiler.record_function'> will be ignored
I0813 11:03:17.741000 1005502 torch/_functorch/_aot_autograd/autograd_cache.py:479] [7/0] AOTAutograd cache miss for key az7cdmpzlztluqvtalqyenr5i4gtqyiobfkiqxcke5tx5tejpo2g
I0813 11:03:32.728000 1005502 torch/_functorch/_aot_autograd/autograd_cache.py:558] [7/0] Writing AOTAutograd cache entry to /tmp/tmpxrdfdu4m/aotautograd/az7cdmpzlztluqvtalqyenr5i4gtqyiobfkiqxcke5tx5tejpo2g/entry
I0813 11:03:33.019000 1005502 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 11:03:33.023000 1005502 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 2
V0813 11:03:33.043000 1005502 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 0 of graph recording id 0
V0813 11:03:33.881000 1005502 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 1 of graph recording id 1
V0813 11:03:34.623000 1005502 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 2 of graph recording id 2
running benchmark: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 27.04it/s]
4.423x
Performing warm-start run for nanogpt
loading model: 0it [00:00, ?it/s]number of parameters: 123.69M
num decayed parameter tensors: 50, with 124,354,560 parameters
num non-decayed parameter tensors: 98, with 121,344 parameters
using fused AdamW: True
loading model: 0it [00:03, ?it/s]
cuda train nanogpt
I0813 11:03:48.888000 1015397 torch/_functorch/_aot_autograd/autograd_cache.py:474] [2/0_1] AOTAutograd cache hit for key adctpnaix4qr2ofjsrnfduh5v4m7qsmnnuy532v45igmy2cplgyu
I0813 11:03:49.133000 1015397 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 11:03:49.270000 1015397 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 0
I0813 11:03:49.295000 1015397 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 11:03:49.296000 1015397 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 1
W0813 11:03:49.309000 1015397 torch/_logging/_internal.py:1058] [5/0] Profiler function <class 'torch.autograd.profiler.record_function'> will be ignored
I0813 11:03:51.532000 1015397 torch/_functorch/_aot_autograd/autograd_cache.py:474] [7/0] AOTAutograd cache hit for key az7cdmpzlztluqvtalqyenr5i4gtqyiobfkiqxcke5tx5tejpo2g
I0813 11:03:51.799000 1015397 torch/_inductor/cudagraph_trees.py:385] [__cudagraphs] recording cudagraph tree for graph without symints
V0813 11:03:51.803000 1015397 torch/_inductor/cudagraph_trees.py:2160] [__cudagraphs] Running warmup of function 2
V0813 11:03:51.820000 1015397 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 0 of graph recording id 0
V0813 11:03:51.986000 1015397 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 1 of graph recording id 1
V0813 11:03:52.185000 1015397 torch/_inductor/cudagraph_trees.py:2119] [__cudagraphs] Recording function 2 of graph recording id 2
running benchmark: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [00:00<00:00, 43.34it/s]
3.021x
```

Cache miss tlparse:
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/jjwu/custom2/dedicated_log_torch_trace_9ew19moa.log/index.html


Cache hit tlparse: 
https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/jjwu/custom2/dedicated_log_torch_trace_0okl70vc.log/index.html

You can click on Perfetto to get the perfetto link for each of the tlparses, here are images for non meta folks:
image
image

> How can I tell from that tlparse that we had an AOT cache hit / cache miss (due to cudagraphs)?

You can see the cache hits and cache misses by mousing over the little arrows in the perfetto link. Note that cudagraphs doesn't affect cache hits or misses since it's just a post compile step. The unit tests do showcase some examples of when cudagraphs itself gets disabled, due to it being disabled on the original cached artifact, though this does not happen with our benchmarks, since our benchmarks naturally don't want to disable cudagraphs.

**Review by eellison — APPROVED** [review]

File: torch/_functorch/aot_autograd.py [hunk 1]
```diff
@@ -941,6 +942,7 @@ def aot_module_simplified(
     decompositions: Optional[Dict] = None,
     keep_inference_input_mutations=False,
     inference_compiler: Optional[Callable] = None,
+    cudagraphs: Optional[BoxedBool] = None,
```
[the remark this answers is not in the document]
Comment (jamesjwu) [reply torch/_functorch/aot_autograd.py:945]: I think you're right, `aot_module_simplified` is only ever called from compile_fx so we should always be able to pass it in, will fix

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 2]
```diff
@@ -165,6 +165,7 @@ def __init__(
         gm: torch.fx.GraphModule,
         example_inputs,
         aot_config: AOTConfig,
+        fx_config: Dict[str, BoxedBool],
```
[the remark this answers is not in the document]
Comment (jamesjwu) [reply torch/_functorch/_aot_autograd/autograd_cache.py:168]: I think it's fully possible that we need other configs from compile_fx in the future, so I'm future proofing it now. I could make a full dataclass for it too but considering the FXGraphCache one is also just a dictionary of values I didn't want to overengineer *too* much, so the dict was a compromise. Should be easy to refactor later if we want to make it a dataclass.

File: torch/_functorch/aot_autograd.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamesjwu) [reply torch/_functorch/aot_autograd.py:945]: Update so, the options here are to make it a *non* default argument, which means every single test/use case needs to pass it in, which seems like a bunch of churn for no reason, or have it be a default argument. But if it's a default argument, having it be a mutable default argument in a boxedbool seems sus. Instead, I'll add a line after this that sets cudagraphs to BoxedBool(config.triton.cudagraphs) if it's not passed in, just like FXGraphCache.

**Pull request closed without merging**